### PR TITLE
Monitor comm cost

### PR DESCRIPTION
**Pull request by yy462**
add communication monitoring during initialization in NC, GC, LP

### CODE REVIEW

File: fedgraph/federated_methods.py [hunk 1]
```diff
@@ -1049,10 +1085,14 @@ def __init__(self, *args, **kwargs):  # type: ignore
         hidden_channels=hidden_channels,
     )
 
+    if args.use_cluster:
+        monitor = monitor or Monitor()
+        monitor.init_time_end()
+
```
Comment (Ryan-YuanLi) [inline fedgraph/federated_methods.py:1091]: monitor should only been initialized at same place, if not monitor throw an error and print error info

File: fedgraph/federated_methods.py [hunk 2]
```diff
@@ -1012,7 +1045,10 @@ def __init__(self, *args, **kwargs):  # type: ignore
 
     current_dir = os.path.dirname(os.path.abspath(__file__))
     ray.init()
-    # Append paths relative to the current script's directory
+    # Use monitor passed from run_fedgraph
+    if args.use_cluster:
+        monitor = args.monitor
```
Comment (Ryan-YuanLi) [inline fedgraph/federated_methods.py:1050]: monitor is an instance not a args

File: fedgraph/federated_methods.py [hunk 3]
```diff
@@ -50,19 +50,38 @@ def run_fedgraph(args: attridict) -> None:
...
+
     if args.fedgraph_task == "NC":
         run_NC(args, data)
     elif args.fedgraph_task == "GC":
         run_GC(args, data)
     elif args.fedgraph_task == "LP":
         run_LP(args)
 
+    # End total communication timing
+    monitor.total_comm_time_end()
```
Comment (Ryan-YuanLi) [inline fedgraph/federated_methods.py:83]: no idea why we need total_comm_time

File: fedgraph/federated_methods.py [hunk 4]
```diff
@@ -50,19 +50,38 @@ def run_fedgraph(args: attridict) -> None:
         Input data for the federated learning task. Format depends on the specific task and
         will be explained in more detail below inside specific functions.
     """
+    # Track initialization and total communication time
+    monitor = Monitor()
+    monitor.init_time_start()
+    monitor.total_comm_time_start()
+
+    args.monitor = monitor
```
Comment (Ryan-YuanLi) [inline fedgraph/federated_methods.py:58]: dont do this way to pass with args

File: distributed_job.py [hunk 5]
```diff
@@ -0,0 +1,13 @@
+import ray
+
+ray.init(address="auto")
+
+
```
Comment (Ryan-YuanLi) [inline distributed_job.py:5]: delete this file

File: fedgraph/federated_methods.py [hunk 6]
```diff
@@ -674,7 +705,8 @@ def run_GC_Fed_algorithm(
         Pandas dataframe with test accuracies
     """
     if server.use_cluster:
-        monitor = Monitor()
+        # Use monitor from server
+        monitor = server.monitor if hasattr(server, "monitor") else Monitor()
```
Comment (Ryan-YuanLi) [inline fedgraph/federated_methods.py:709]: only initialize at one place

**Review by Ryan-YuanLi — CHANGES_REQUESTED** [review]

File: quickstart.py [hunk 7]
```diff
@@ -58,96 +59,96 @@
...
-    # Output configuration
-    "outbase": "./outputs",
-    "save_files": False,
-    # Scalability and Cluster Configuration
-    "use_cluster": False,  # Use Kubernetes for scalability if True
-}
-#######################################################################
-# Run fedgraph method
-# -------------------
+# #######################################################################
```
[the remark this answers is not in the document]
Comment (yy462) [reply quickstart.py:62]: Thanks for remind! Cause last time, to save time, when I tried to do experiment in local, I comment out it. Gonna make it back!

File: fedgraph/federated_methods.py (hunk 1 above)
Comment (yy462) [reply fedgraph/federated_methods.py:1091]: Yea, it is correct, the reason why I do this is that I wanted to ensure monitor is always defined when use_cluster is true, in case it’s not passed correctly. 
Gonna throw an error and print error info.

File: fedgraph/federated_methods.py (hunk 2 above)
Comment (yy462) [reply fedgraph/federated_methods.py:1050]: The reason why I tried this, cause if we round the project in local, there will be some error when we pass monitor, and I also found we repeatly initial Monitor in some parts of our original code. Like In the helper functions such as run_GC_selftrain, run_GC_Fed_algorithm and run_GCFL_algorithm, we still create a new Monitor instance each time. Should we need to change this

File: fedgraph/federated_methods.py (hunk 3 above)
Comment (yy462) [reply fedgraph/federated_methods.py:83]: OK, already delete all the code that related total_comm_time. I think I misunderstand what you guys required. For now, I just calculate each initialization time for each run_NC, run_GC, run_LP.

File: fedgraph/federated_methods.py (hunk 4 above)
Comment (yy462) [reply fedgraph/federated_methods.py:58]: Just delete it and change it back

File: distributed_job.py (hunk 5 above)
Comment (yy462) [reply distributed_job.py:5]: Already deleted.

File: fedgraph/federated_methods.py (hunk 6 above)
Comment (yy462) [reply fedgraph/federated_methods.py:709]: Goi it!

**Review by Ryan-YuanLi — APPROVED** [review]

File: fedgraph/federated_methods.py [hunk 8]
```diff
@@ -50,21 +50,26 @@ def run_fedgraph(args: attridict) -> None:
...
         data = None
     if args.fedgraph_task == "NC":
-        run_NC(args, data)
+        run_NC(args, data, monitor)
     elif args.fedgraph_task == "GC":
-        run_GC(args, data)
+        run_GC(args, data, monitor)
     elif args.fedgraph_task == "LP":
-        run_LP(args)
+        run_LP(args, monitor)
```
Comment (Ryan-YuanLi) [inline fedgraph/federated_methods.py:68]: remove this, always init inside of the function

File: fedgraph/federated_methods.py [hunk 9]
```diff
@@ -301,9 +303,9 @@ def __init__(self, *args: Any, **kwds: Any):
                 server.trainers[i].load_feature_aggregation.remote(trainer_aggregation)
             print("clients received feature aggregation from server")
         [trainer.relabel_adj.remote() for trainer in server.trainers]
-    if args.use_cluster:
-        monitor.pretrain_time_end(30)
-        monitor.train_time_start()
+
+    monitor.pretrain_time_end(30)
+    monitor.train_time_start()
     #######################################################################
```
Comment (Ryan-YuanLi) [inline fedgraph/federated_methods.py:309]: we could only sleep 30s when use_cluster

File: fedgraph/federated_methods.py [hunk 10]
```diff
@@ -740,7 +748,7 @@ def highlight_max(s: pd.Series) -> list:
         is_max = s == s.max()
         return ["background-color: yellow" if v else "" for v in is_max]
 
-    if server.use_cluster:
+    if monitor is not None:
         monitor.train_time_end(30)
     fs = frame.style.apply(highlight_max).data
```
Comment (Ryan-YuanLi) [inline fedgraph/federated_methods.py:753]: init inside the function so it wont be none, too many if

File: fedgraph/federated_methods.py [hunk 11]
```diff
@@ -1023,10 +1031,9 @@ def __init__(self, *args, **kwargs):  # type: ignore
 
     current_dir = os.path.dirname(os.path.abspath(__file__))
     ray.init()
-    if args.use_cluster:
-        # Initialize monitor and start tracking initialization time
-        monitor = Monitor()
+    if args.use_cluster and monitor is not None:
         monitor.init_time_start()
+
     # Append paths relative to the current script's directory
```
Comment (Ryan-YuanLi) [inline fedgraph/federated_methods.py:1037]: same issue, still use args.use_cluster

File: fedgraph/federated_methods.py [hunk 12]
```diff
@@ -304,7 +300,7 @@ def __init__(self, *args: Any, **kwds: Any):
             print("clients received feature aggregation from server")
         [trainer.relabel_adj.remote() for trainer in server.trainers]
 
-    monitor.pretrain_time_end(30)
+    monitor.pretrain_time_end(30 if args.use_cluster else 0)
     monitor.train_time_start()
```
Comment (Ryan-YuanLi) [inline fedgraph/federated_methods.py:304]: too hack, could you do this if else inside of the monitor

**Review by Ryan-YuanLi — CHANGES_REQUESTED** [review]

File: fedgraph/federated_methods.py (hunk 8 above)
Comment (yy462) [reply fedgraph/federated_methods.py]: Yea, that make sense, thanks for the advice.

**Review by Ryan-YuanLi — APPROVED** [review]

File: ray_cluster_configs/eks_cluster_config.yaml.bak [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+    maxSize: 1
+    volumeSize: 256
+    labels:
+      ray-node-type: head
+
+  - name: worker-nodes
+    instanceType: m5.16xlarge
+    desiredCapacity: 10
+    minSize: 10
+    maxSize: 10
```
Comment (Ryan-YuanLi) [inline ray_cluster_configs/eks_cluster_config.yaml.bak:22]: why there's a bak

File: setup_cluster.sh [hunk 14]
```diff
@@ -113,39 +116,45 @@ echo "If any pod status is Pending, modify ray_kubernetes_cluster.yaml and reapp
...
-  \"working_dir\": \"./\",
-  \"excludes\": [\".git\"]
-}' --address http://localhost:8265 -- python3 run.py"
+echo "ray job submit \
+  --address http://localhost:8265 \
+  --runtime-env-json '{
+    "working_dir": ".",
+    "excludes": [".git", "__pycache__", "outputs", "fedgraph/he_training_context.pkl"],
+    "pip": ["fsspec==2023.6.0", "huggingface_hub", "tenseal"]
+  }' \
```
Comment (Ryan-YuanLi) [inline setup_cluster.sh:145]: "fsspec==2023.6.0" may be too strict

File: fedgraph/federated_methods.py [hunk 15]
```diff
@@ -144,7 +144,13 @@ def run_NC(args: attridict, data: Any = None) -> None:
     class Trainer(Trainer_General):
         def __init__(self, *args: Any, **kwds: Any):
             super().__init__(*args, **kwds)
-            self.use_encryption = kwds["args"].use_encryption
+            args = kwds.get("args", {})
+            self.use_encryption = (
+                getattr(args, "use_encryption", False)
+                if hasattr(args, "use_encryption")
```
[the remark this answers is not in the document]
Comment (yy462) [reply fedgraph/federated_methods.py:150]: Oh, thanks for the remind. I've renamed the local variable to args_obj to avoid shadowing *args, ensuring clarity and safety in this part.

File: ray_cluster_configs/eks_cluster_config.yaml.bak (hunk 13 above)
Comment (yy462) [reply ray_cluster_configs/eks_cluster_config.yaml.bak:22]: The .bak file is a backup automatically created by Ray when updating the EKS cluster config to prevent accidental loss. It’s safe to ignore or delete if version control is used.

File: setup_cluster.sh (hunk 14 above)
Comment (yy462) [reply setup_cluster.sh:145]: Yea, I think we can relax the version constraint to just "fsspec" to avoid potential compatibility issues.

**Review by Ryan-YuanLi — APPROVED** [review]
approve

File: fedgraph/federated_methods.py [hunk 16]
```diff
@@ -1035,6 +1060,9 @@ def setup_trainer_server(
         class Trainer(Trainer_LP):
             def __init__(self, *args, **kwargs):  # type: ignore
                 super().__init__(*args, **kwargs)
+                print(
+                    f"[Debug] Trainer running on node IP: {ray.util.get_node_ip_address()}"
+                )
 
         clients = [
```
[the remark this answers is not in the document]
Comment (yy462) [reply fedgraph/federated_methods.py:1067]: Yes, it’s possible — if trainers’ resource demands are small and Ray schedules multiple trainers onto the same node or pod, they will share the same IP.
